### DISCUSSION TITLE
trigger change in element through button

### DISCUSSION THREAD

**Discussion by Scheeer** [zauberzeug/nicegui#1318] [Q&A] [answered]
### Question

Hi, 
im trying to increase a slider value via button click.
Every time the slider is changed something is supposed to happen, in the case of this minimal example a notification with success is sent. 

When pressing the buttons the slider value changes, the slider is updated, but the on_change event is not called. Im not sure how to proceed here. I also tried to edit _props variable and calling measurements.update.

```python
from nicegui import ui

def increase_slider():
    if measurements.value > int(measurements._props["max"]):
        measurements.value = int(measurements._props["max"])
    else:
        measurements.value = measurements.value+1
    ui.notify("test")

def decrease_slider():
    if measurements.value-1 < int(measurements._props["min"]):
        measurements.value = int(measurements._props["min"])
    else:
        measurements.value = measurements.value-1
    ui.notify("test")
        
prev_measurement_button = ui.button(icon='arrow_circle_left',on_click = decrease_slider)
next_measurement_button = ui.button(icon="arrow_circle_right",on_click = increase_slider)
measurements = ui.slider(min=0, max=100, value=50,step=1).props('label').on('change', lambda: ui.notify("success"),                                                                                                               
throttle=1.0, leading_events=False)
runmode = ui.run(dark=True,reload = False,native = True)
```

**Comment by falkoschindler** [comment]
Hi @Scheeer,
That's an interesting question. I don't immediately see the reason why the change event isn't triggered.
Just for the record, here is a minimum reproduction:
```py
ui.button('<', on_click=lambda: slider.set_value(max(slider._props['min'], slider.value - 1)))
ui.button('>', on_click=lambda: slider.set_value(min(slider._props['max'], slider.value + 1)))
slider = ui.slider(min=0, max=10, value=5).on('change', lambda: ui.notify('slider value changed'))
```
Clicking the buttons should trigger a notification, but it doesn't.

**Comment by falkoschindler** [comment]
I think I understand: Even though `.on()` registers an event listener, a value change only triggers the `on_change` handler passed to the slider like this:

```py
slider = ui.slider(min=0, max=10, value=5, on_change=lambda: ui.notify('slider value changed'))
```

Is there a specific reason why you chose to use `.on()`? Do you need the `toggle` and `leading_events` argument?

We could consider all event listeners registered with `.on()` for handling value changes:
```py
    def on_value_change(self, value: Any) -> None:
        ...
        for event_listener in self._event_listeners.values():
            if event_listener.type == 'change':
                handle_event(event_listener.handler, args)
```
But it is not trivial to decide which types of events should be triggered: Should a value change trigger "change", "update:model-value", or both? Usually these events originate from user input on the client, therefore they don't match 1:1 to a value change caused by some Python logic.

**Reply by Scheeer** [reply]
yea, i chose `.on()` because it is not desired to permanently trigger the event when changing the slider by user input

**Comment by falkoschindler** [comment, marked as answer]
I think the easiest solution is to explicitly call the event handler, when the event is triggered by the client _and_ when it is triggered by changing the value via button click:

```py
def increase_slider():
    slider.value = min(slider._props['max'], slider.value + 1)
    handle_change()

def decrease_slider():
    slider.value = max(slider._props['min'], slider.value - 1)
    handle_change()

def handle_change():
    ui.notify('slider value changed')

ui.button('<', on_click=decrease_slider)
ui.button('>', on_click=increase_slider)
slider = ui.slider(min=0, max=10, value=5).on('change', handle_change, throttle=1.0, leading_events=False)
```

**Reply by Scheeer** [reply]
its not very elegant, but it does work. I don't know enough about the differences of "change" and "update:model-value" to say which one should be triggered by python logic. I do however feel like at least change should trigger it as i was under the impression that any change to this element would trigger this event.
